unixfs/hamt/fetcher.go

```diff
@@ -0,0 +1,321 @@
+	resultCh  chan result // channel for retrieving the results of the request
+
+	idle bool
+
+	done chan batchJob // when a batch job completes results are sent to this channel
+
+	todoFirst *job            // do this job first since we are waiting for its results
+	todo      jobStack        // stack of jobs that still need to be done
+	jobs      map[*Shard]*job // map of all jobs in which the results have not been collected yet
+
+	// stats relevant for streaming the complete hamt directory
+	doneCnt    int // job's done but results not yet retrieved
+	hits       int // job's result already ready, no delay
+	nearMisses int // job currently being worked on, small delay
+	misses     int // job on todo stack but will be done in the next batch, larger delay
+
+	// other useful stats
+	cidCnt int
+
+	start time.Time
+}
+
+// batchSize must be at least as large as the largest number of CIDs
+// requested in a single job. For best performance it should likely be
+// slightly larger as jobs are popped from the todo stack in order and a
+// job close to the batchSize could force a very small batch to run.
+// The recommend minimum size is thus a size slightly larger than the
+// maximum number children in a HAMT node (which is the largest number
+// of CIDs that could be requested in a single job) or 256 + 64 = 320
+const batchSize = 320
+
+//
+// fetcher public interface
+//
+
+// startFetcher starts a new fetcher in the background
+func startFetcher(ctx context.Context, dserv ipld.DAGService) *fetcher {
+	log.Infof("fetcher: starting...")
+	f := &fetcher{
+		ctx:       ctx,
+		dserv:     dserv,
+		requestCh: make(chan *Shard),
+		resultCh:  make(chan result),
+		idle:      true,
+		done:      make(chan batchJob),
+		jobs:      make(map[*Shard]*job),
+	}
+	go f.mainLoop()
+	return f
+}
+
+// result contains the result of a job, see getResult
+type result struct {
+	vals map[string]*Shard
+	errs []error
+}
+
+// get gets the missing child shards for the hamt object.
+// The missing children for the passed in shard are returned.  The
+// children are then also retrieved in the background.  The result is
+// the result of the batch request and not just the single job.  In
+// particular, if the 'errs' field is empty the 'vals' of the result
+// is guaranteed to contain all the missing child shards, but the
+// map may also contain child shards of other jobs in the batch.
+func (f *fetcher) get(hamt *Shard) result {
+	f.requestCh <- hamt
+	res := <-f.resultCh
+	return res
+}
+
+//
+// fetcher internals
+//
+
+type job struct {
+	id  *Shard
+	idx int /* index in the todo stack, an index of -1 means the job
+	   is already done or being worked on now */
+	cids []*cid.Cid
+	res  result
+}
+
+func (f *fetcher) mainLoop() {
+	var want *Shard /* want is the job id (the parent shard) of the
+	   results we want next */
+	f.start = time.Now()
+	for {
+		select {
+		case id := <-f.requestCh:
+			if want != nil {
```

Comment:
seems like we could just pull most of the code in this select case into a separate function, making this flow easier to read and reason about (also removing the need for continues)

Reply:
Okay.  I agree this will be helpful.  I will do this later today.

Reply:
Okay done.